### PR TITLE
PWN-7765 - Add History pending transactions

### PR DESCRIPTION
**Pull request by Davran331122**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-7765

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -27,7 +27,7 @@
         android:backupAgent="AppPrefsBackupAgent"
         android:dataExtractionRules="@xml/data_backup_rules"
         android:fullBackupContent="@xml/backup_rules"
-        android:largeHeap="false"
+        android:largeHeap="true"
```
Comment (gslevinkov) [inline app/src/main/AndroidManifest.xml:30]: Damn boy, what's with the emulator? Install some more ram :D

File: app/src/main/java/org/p2p/wallet/history/repository/local/TransactionDetailsDatabaseRepository.kt [hunk 2]
```diff
@@ -1,18 +1,37 @@
...
 import org.p2p.solanaj.kits.transaction.TransactionDetails
+import org.p2p.wallet.history.model.HistoryTransaction
 import org.p2p.wallet.history.repository.local.mapper.TransactionDetailsEntityMapper
 import org.p2p.wallet.history.repository.local.db.dao.TransactionsDaoDelegate
 
 class TransactionDetailsDatabaseRepository(
     private val daoDelegate: TransactionsDaoDelegate,
     private val mapper: TransactionDetailsEntityMapper,
 ) : TransactionDetailsLocalRepository {
+    private val pendingTransactions = HashMap<String, HistoryTransaction?>()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/local/TransactionDetailsDatabaseRepository.kt:12]: Just mutable map doesn't work?

File: app/src/main/java/org/p2p/wallet/history/repository/local/TransactionDetailsDatabaseRepository.kt [hunk 3]
```diff
@@ -1,18 +1,37 @@
...
+
     override suspend fun getTransactions(signatures: List<String>): List<TransactionDetails> {
         return daoDelegate.getTransactions(signatures)
             .let { mapper.fromEntityToDomain(it) }
     }
 
+    override suspend fun savePendingTransaction(txSignature: String, transaction: HistoryTransaction) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/local/TransactionDetailsDatabaseRepository.kt:19]: It's not database repository anymore :( 
InMemory - yes

File: app/src/main/java/org/p2p/wallet/history/repository/remote/HistoryRepository.kt [hunk 4]
```diff
@@ -59,7 +61,14 @@ class HistoryRepository(
             if (errorMessage.isNotEmpty()) {
                 return@withContext HistoryPagingResult.Error(Throwable(errorMessage))
             }
-            return@withContext newTransactions.sortedByDescending { transaction ->
+            val newTransactionIds = newTransactions.map { it.getHistoryTransactionId() }
+            localRepository.getAllPendingTransactions().forEach {
+                if (it.getHistoryTransactionId() in newTransactionIds)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/history/repository/remote/HistoryRepository.kt:66]: {}

File: app/src/main/java/org/p2p/wallet/jupiter/ui/main/JupiterSwapPresenter.kt [hunk 5]
```diff
@@ -181,6 +188,24 @@ class JupiterSwapPresenter(
                     stateManager.onNewAction(SwapStateAction.CancelSwapLoading)
                     val transactionState = TransactionState.JupiterSwapSuccess
                     transactionManager.emitTransactionState(internalTransactionId, transactionState)
+
+                    val pendingTransaction = RpcHistoryTransaction.Swap(
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/jupiter/ui/main/JupiterSwapPresenter.kt:192]: Maybe extract this big constructor to companion object and pass only currentState and result?

File: app/src/main/java/org/p2p/wallet/jupiter/ui/main/JupiterSwapPresenter.kt [hunk 6]
```diff
@@ -206,7 +231,8 @@ class JupiterSwapPresenter(
             is SwapState.TokenANotZero,
             is SwapState.LoadingRoutes,
             is SwapState.RoutesLoaded,
-            is SwapState.LoadingTransaction -> swapInteractor.getTokenAAmount(featureState)
+            is SwapState.LoadingTransaction,
+            -> swapInteractor.getTokenAAmount(featureState)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/jupiter/ui/main/JupiterSwapPresenter.kt:235]: Strange line break

**Review by kabramovp2p — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/history/repository/local/TransactionDetailsDatabaseRepository.kt (hunk 3 above)
Comment (Davran331122) [reply app/src/main/java/org/p2p/wallet/history/repository/local/TransactionDetailsDatabaseRepository.kt:19]: okay, but there is dao delegate )

**Review by NikolaiEmelianov — APPROVED** [review]

**Pull request merged**